packages/react-dom/src/__tests__/ReactMount-test.js

```diff
@@ -173,12 +482,632 @@ describe('ReactMount', () => {
+    // See fixtures/ssr: ssr-warnForDeletedHydratableInstance-didNotHydrateInstance
+
+    const div = document.createElement('div');
+    const markup = ReactDOMServer.renderToString(
+      <div>
+        SSRMismatchTest server text
+        <span />
+        <div data-ssr-mismatch-padding-after="1" />
+        <div data-ssr-mismatch-padding-after="2" />
+        <div data-ssr-mismatch-padding-after="3" />
+        <div data-ssr-mismatch-padding-after="4" />
+        <div data-ssr-mismatch-padding-after="5" />
+      </div>,
+    );
+    div.innerHTML = markup;
+
+    expect(() =>
+      ReactDOM.hydrate(
+        <div>
+          <span />
+        </div>,
+        div,
+      ),
+    ).toWarnDev(
+      "Warning: Did not expect server HTML to contain the text node {'SSRMismatchTest server text'} in <div>.\n\n" +
+        '  <div data-reactroot="">\n' +
+        "-   {'SSRMismatchTest server text'}\n" +
+        '    <span />\n' +
+        '    <div data-ssr-mismatch-padding-after="1" />\n' +
+        '    <div data-ssr-mismatch-padding-after="2" />\n' +
+        '    <div data-ssr-mismatch-padding-after="3" />\n' +
+        '    <div data-ssr-mismatch-padding-after="4" />\n' +
+        '    <div data-ssr-mismatch-padding-after="5" />\n' +
+        '  </div>\n\n' +
+        '    in span (at **)\n' +
+        '    in div (at **)',
+    );
+  });
+
+  it('should warn when hydrate replaces a text node by an element within a root container (replacement diff)', () => {
+    // See fixtures/ssr: ssr-warnForInsertedHydratedInstance-didNotFindHydratableContainerInstance
+
+    const div = document.createElement('div');
+    const markup = 'SSRMismatchTest default text';
+    div.innerHTML = markup;
+
+    expect(() =>
+      ReactDOM.hydrate(<span>SSRMismatchTest default text</span>, div),
+    ).toWarnDev(
+      'Warning: Expected server HTML to contain a matching <span> in <div>.\n\n' +
+        '  <div>\n' +
+        "-   {'SSRMismatchTest default text'}\n" +
+        '+   <span>SSRMismatchTest default text</span>\n' +
+        '  </div>\n\n' +
+        '    in span (at **)',
+    );
+  });
+
+  it('should warn when hydrate replaces an element by a text node within a root container (replacement diff)', () => {
+    // See fixtures/ssr: ssr-warnForInsertedHydratedTextInstance-didNotFindHydratableContainerTextInstance
+
+    const div = document.createElement('div');
+    const markup = ReactDOMServer.renderToString(
+      <span>SSRMismatchTest default text</span>,
+    );
+    div.innerHTML = markup;
+
+    expect(() =>
+      ReactDOM.hydrate('SSRMismatchTest default text', div),
+    ).toWarnDev(
+      'Warning: Expected server HTML to contain a matching text node' +
+        " for {'SSRMismatchTest default text'} in <div>.\n\n" +
+        '  <div>\n' +
+        '-   <span data-reactroot="">SSRMismatchTest default text</span>\n' +
+        "+   {'SSRMismatchTest default text'}\n" +
```

Comment:
I decided to always print the top-level plain text children as strings in curly braces, but this can be changed with a single boolean flip, see `printElementOrTextForHydrationDiff`.

Reply:
Refactored this flag to be an enum value, `PRINT_CHILDREN_VALUE_MODE_DEFAULT` or `PRINT_CHILDREN_VALUE_MODE_RAW` (see `getHydrationDiff`).